plenum/server/consensus/view_change_service.py

```diff
@@ -133,29 +137,65 @@ def __init__(self, data: ConsensusSharedData, timer: TimerService, bus: Internal
         self._timer = timer
         self._bus = bus
         self._network = network
-        self._router = StashingRouter(self._config.VIEW_CHANGE_SERVICE_STASH_LIMIT)
+        self._router = stasher
         self._votes = ViewChangeVotesForView(self._data.quorums)
         self._new_view = None  # type: Optional[NewView]
 
         self._router.subscribe(ViewChange, self.process_view_change_message)
         self._router.subscribe(ViewChangeAck, self.process_view_change_ack_message)
         self._router.subscribe(NewView, self.process_new_view_message)
-        self._router.subscribe_to(network)
 
         self._old_prepared = {}  # type: Dict[int, BatchID]
         self._old_preprepared = {}  # type: Dict[int, List[BatchID]]
         self._primaries_selector = RoundRobinPrimariesSelector()
 
+        self._bus.subscribe(NeedViewChange, self.process_need_view_change)
+
     def __repr__(self):
         return self._data.name
 
-    def start_view_change(self, view_no: Optional[int] = None):
+    def process_need_view_change(self, msg: NeedViewChange):
```

Comment:
Should we check, that view_change already in progress or msg.view_no is not the next view_no? For example, now view_no is 1 but msg.view_no is 3 or greater?

Reply:
We may run this even if view change is in progress if we were not able to finish it last round (because the primary was not connected for example).
As for jumping to more than 1 view, I think in theory it's possible for lagging node...
I think we can add additional validation here if necessary during integration phase.